src/nagini_translation/analyzer.py

```diff
@@ -460,15 +460,53 @@ def visit_ClassDef(self, node: ast.ClassDef) -> None:
                 if not (isinstance(base, ast.Name) and
                         base.id in MYPY_SUPERCLASSES):
                     actual_bases.append(base)
+
+        if len(actual_bases) > 0:
+            cls.superclass = self.find_or_create_target_class(actual_bases[0])
+
+            # Process ADT defined by a class hierarchy (sum) and a NamedTuple (product)
+            if isinstance(cls.superclass, PythonClass) and cls.superclass.is_adt:
+                cls.is_adt = True
+                # Detect malformed ADTs
+                # ADT classes can have at most two superclasses
+                if len(actual_bases) > 2:
```

Comment:
Could you put this whole ADT well-formedness checking in a separate method?

Reply:
Yes, I also created another method to visit the ADT which calls the newly created method to check if the ADT is well-formed (matching better the existing code structure).